mautic/mautic
# DPMMA-2608 Local Autosave and Recovery Feature for GrapesJS Builder

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ N ]
| New feature/enhancement? (use the a.x branch)      | [ Y ]
| Deprecations?                          | [ N ]
| BC breaks? (use the c.x branch)        | [ N ]
| Automated tests included?              | [ N ] (only JS and CSS changes)
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:
This pull request introduces a local autosave backup and recovery functionality for the GrapesJS builder used in Mautic for emails and pages.

Features:
- Functionality to store a local copy of the builder content in the user's browser local storage each time the content is modified.
- When an email or page is accessed in Mautic, the editor will check if a different version exists in the browser's local storage. If a different version is found, a prompt will be displayed to the user, giving them the option to restore the autosaved version. If the user chooses to restore, the contents of the editor will be replaced with the autosaved version from the local storage.
- The copy is associated with the type (page/email), theme (e.g. blank/paprika) and the entity ID
- The storage can contain last 10 copies to prevent web storage from exceeding its 10MiB per domain limit.



#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Open the email or page builder, either by using an existing object or creating a new one.
3. Choose a theme for the page or email.
4. Open the builder and make some changes.
5. Close the browser tab without saving any changes.
6. Reopen the builder, ensuring that the type (page/email), theme, and ID match the previous object or a new one.
7. Verify that the copy is available and click the "Restore the backup" button.
![image](https://github.com/mautic/mautic/assets/8580942/d2dabb52-c916-45c5-aae4-59ec760b1a15)
8. Confirm that the content has been successfully restored.



This PR includes the commit [6486428e389d6626398558e05c512db4a1fe522b](https://github.com/mautic/mautic/commit/6486428e389d6626398558e05c512db4a1fe522b) from the 5.0 branch to prevent future conflicts.
